File: cli/src/katello/client/api/system.py
# -*- coding: utf-8 -*-
#
# Copyright © 2012 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
#
# Red Hat trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate Red Hat trademarks that are incorporated
# in this software or its documentation.

from katello.client.api.base import KatelloAPI
from katello.client.utils.encoding import u_str

# pylint: disable=R0904
class SystemAPI(KatelloAPI):
    """
    Connection class to access environment calls
    """
    def register(self, name, org, environment_id, activation_keys, cp_type, release=None, sla=None, facts=None):
        if environment_id is not None:
            path = "/api/environments/%s/systems" % environment_id
        else:
            path = "/api/organizations/%s/systems" % org
        facts_with_defaults = { "distribution.name": "Unknown", "cpu.cpu_socket(s)": "1" }
        if facts is not None:
            facts_with_defaults.update(facts)
        sysdata = {
            "name": name,
            "cp_type": cp_type,
            "serviceLevel": sla,
            "facts": facts_with_defaults
        }
        if activation_keys:
            sysdata["activation_keys"] = activation_keys
        if release:
            sysdata["releaseVer"] = release

        return self.server.POST(path, sysdata)[1]

    def unregister(self, system_uuid):
        path = "/api/systems/" + u_str(system_uuid)
        return self.server.DELETE(path)[1]

    def subscribe(self, system_id, pool, quantity):
        path = "/api/systems/%s/subscriptions" % system_id
        data = {
            "pool": pool,
            "quantity": quantity
        }
        return self.server.POST(path, data)[1]

    def subscriptions(self, system_id):
        path = "/api/systems/%s/subscriptions" % system_id
        return self.server.GET(path)[1]

    def available_pools(self, system_id, match_system=False, match_installed=False, no_overlap=False):
        params = {}
        self.update_dict(params, "match_system", match_system)
        self.update_dict(params, "match_installed", match_installed)
        self.update_dict(params, "no_overlap", no_overlap)

        path = "/api/systems/%s/pools" % system_id

        return self.server.GET(path, params)[1]

    def unsubscribe(self, system_id, entitlement):
        path = "/api/systems/%s/subscriptions/%s" % (system_id, entitlement)
        return self.server.DELETE(path)[1]

    def unsubscribe_by_serial(self, system_id, serial):
        path = "/api/systems/%s/subscriptions/serials/%s" % (system_id, serial)
        return self.server.DELETE(path)[1]

    def unsubscribe_all(self, system_id):
        path = "/api/systems/%s/subscriptions/" % system_id
        return self.server.DELETE(path)[1]

    def system(self, system_id):
        path = "/api/systems/%s" % system_id
        return self.server.GET(path)[1]

    def tasks(self, org_name, environment_id, system_name = None, system_uuid = None):
        params = {}
        self.update_dict(params, "environment_id", environment_id)
        if system_name:
            self.update_dict(params, "system_name", system_name)
        if system_uuid:
            self.update_dict(params, "system_uuid", system_uuid)

        path = "/api/organizations/%s/systems/tasks" % org_name
        return self.server.GET(path, params)[1]

    def packages(self, system_id):
        path = "/api/systems/%s/packages" % system_id
        return self.server.GET(path)[1]

    def releases_for_system(self, system_id):
        path = "/api/systems/%s/releases" % system_id
        return self.server.GET(path)[1]

    def releases_for_environment(self, env_id):
        path = "/api/environments/%s/releases" % env_id
        return self.server.GET(path)[1]

    def update(self, system_id, params = None):
        path = "/api/systems/%s" % system_id
        return self.server.PUT(path, params)[1]

    def install_packages(self, system_id, packages):
        path = "/api/systems/%s/packages" % system_id
        return self.server.POST(path, {"packages": packages})[1]

    def remove_packages(self, system_id, packages):
        path = "/api/systems/%s/packages" % system_id
        return self.server.DELETE(path, {"packages": packages})[1]

    def update_packages(self, system_id, packages):
        path = "/api/systems/%s/packages" % system_id
        return self.server.PUT(path, {"packages": packages})[1]

    def install_package_groups(self, system_id, packages):
        path = "/api/systems/%s/packages" % system_id
        return self.server.POST(path, {"groups": packages})[1]

    def remove_package_groups(self, system_id, packages):
        path = "/api/systems/%s/packages" % system_id
        return self.server.DELETE(path, {"groups": packages})[1]

    def systems_by_org(self, orgId, query = None):
        path = "/api/organizations/%s/systems" % orgId
        return self.server.GET(path, query)[1]

    def systems_by_env(self, environment_id, query = None):
        path = "/api/environments/%s/systems" % environment_id
        return self.server.GET(path, query)[1]

    def errata(self, system_id):
        path = "/api/systems/%s/errata" % system_id
        return self.server.GET(path)[1]

    def report_by_org(self, orgId, format_in):
        path = "/api/organizations/%s/systems/report" % orgId
        to_return = self.server.GET(path, custom_headers={"Accept": format_in})
        return (to_return[1], to_return[2])

    def report_by_env(self, env_id, format_in):
        path = "/api/environments/%s/systems/report" % env_id
        to_return = self.server.GET(path, custom_headers={"Accept": format_in})
        return (to_return[1], to_return[2])

    def add_system_groups(self, system_id, system_group_ids):
        data = { 'system' : {
                    'system_group_ids' : system_group_ids
            }
        }
        path = "/api/systems/%s/system_groups/" % system_id
        return self.server.POST(path, data)[1]

    def remove_system_groups(self, system_id, system_group_ids):
        data = { 'system' : {
                    'system_group_ids' : system_group_ids
            }
        }
        path = "/api/systems/%s/system_groups/" % system_id
        return self.server.DELETE(path, data)[1]

    def remove_consumer_deletion_record(self, uuid):
        path = "/api/consumers/%s/deletionrecord" % uuid
        return self.server.DELETE(path)[1]

File: app/models/system.rb

#
# Copyright 2011 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

class NonLibraryEnvironmentValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    return unless value
    record.errors[attribute] << N_("Cannot register a system with 'Library' environment ") if record.environment != nil && record.environment.library?
  end
end

class System < ActiveRecord::Base
  include Glue::Candlepin::Consumer if AppConfig.use_cp
  include Glue::Pulp::Consumer if AppConfig.use_pulp
  include Glue::ElasticSearch::System
  include Glue
  include Authorization::System
  include AsyncOrchestration

  after_rollback :rollback_on_create, :on => :create

  acts_as_reportable

  belongs_to :environment, :class_name => "KTEnvironment", :inverse_of => :systems
  belongs_to :system_template

  has_many :task_statuses, :as => :task_owner, :dependent => :destroy

  has_many :system_activation_keys, :dependent => :destroy
  has_many :activation_keys, :through => :system_activation_keys

  has_many :system_system_groups, :dependent => :destroy
  has_many :system_groups, {:through => :system_system_groups, :before_add => :add_pulp_consumer_group, :before_remove => :remove_pulp_consumer_group}.merge(update_association_indexes)

  has_many :custom_info, :as => :informable, :dependent => :destroy

  validates :environment, :presence => true, :non_library_environment => true
  # multiple systems with a single name are supported
  validates :name, :presence => true, :no_trailing_space => true
  validates :description, :katello_description_format => true
  validates_length_of :location, :maximum => 255
  validates :sockets, :numericality => { :only_integer => true, :greater_than => 0 }, :allow_blank => true, :allow_nil => true, :on => {:create, :update}
  before_create  :fill_defaults

  scope :by_env, lambda { |env| where('environment_id = ?', env) unless env.nil?}
  scope :completer_scope, lambda { |options| readable(options[:organization_id])}

  class << self
    def architectures
      { 'i386' => 'x86', 'ia64' => 'Itanium', 'x86_64' => 'x86_64', 'ppc' => 'PowerPC',
      's390' => 'IBM S/390', 's390x' => 'IBM System z', 'sparc64' => 'SPARC Solaris',
      'i686' => 'i686'}
    end

    def virtualized
      { "physical" => N_("Physical"), "virtualized" => N_("Virtual") }
    end
  end

  def organization
    environment.organization
  end

  def consumed_pool_ids
    self.pools.collect {|t| t['id']}
  end

  def available_releases
    self.environment.available_releases
  end

  def consumed_pool_ids=attributes
    attribs_to_unsub = consumed_pool_ids - attributes
    attribs_to_unsub.each do |id|
      self.unsubscribe id
    end

    attribs_to_sub = attributes - consumed_pool_ids
    attribs_to_sub.each do |id|
      self.subscribe id
    end
  end

  def filtered_pools match_system, match_installed, no_overlap
    pools = self.available_pools !match_system

    # Only available pool's with a product on the system'
    if match_installed
      pools = pools.select do |pool|
        self.installedProducts.any? do |installedProduct|
          pool['providedProducts'].any? do |providedProduct|
            installedProduct['productId'] == providedProduct['productId']
          end
        end
      end
    end

    # None of the available pool's products overlap a consumed pool's products
    if no_overlap
      pools = pools.select do |pool|
        pool['providedProducts'].all? do |providedProduct|
          self.consumed_entitlements.all? do |consumedEntitlement|
            consumedEntitlement.providedProducts.all? do |consumedProduct|
              consumedProduct.cp_id != providedProduct['productId']
            end
          end
        end
      end
    end

    return pools
  end

  def install_packages packages
    pulp_task = self.install_package(packages)
    task_status = save_task_status(pulp_task, :package_install, :packages, packages)
  end

  def uninstall_packages packages
    pulp_task = self.uninstall_package(packages)
    task_status = save_task_status(pulp_task, :package_remove, :packages, packages)
  end

  def update_packages packages=nil
    # if no packages are provided, a full system update will be performed (e.g ''yum update' equivalent)
    pulp_task = self.update_package(packages)
    task_status = save_task_status(pulp_task, :package_update, :packages, packages)
  end

  def install_package_groups groups
    pulp_task = self.install_package_group(groups)
    task_status = save_task_status(pulp_task, :package_group_install, :groups, groups)
  end

  def uninstall_package_groups groups
    pulp_task = self.uninstall_package_group(groups)
    task_status = save_task_status(pulp_task, :package_group_remove, :groups, groups)
  end

  def install_errata errata_ids
    pulp_task = self.install_consumer_errata(errata_ids)
    task_status = save_task_status(pulp_task, :errata_install, :errata_ids, errata_ids)
  end

  def as_json(options)
    json = super(options)
    json['environment'] = environment.as_json unless environment.nil?
    json['activation_key'] = activation_keys.as_json unless activation_keys.nil?
    json['template'] = system_template.as_json unless system_template.nil?
    json['ipv4_address'] = facts.try(:[], 'network.ipv4_address')
    if self.guest == 'true'
      json['host'] = self.host.attributes if self.host
    else
      json['guests'] = self.guests.map(&:attributes)
    end
    json
  end

  def tasks
    TaskStatus.refresh_for_system(self)
  end

  # A rollback occurred while attempting to create the system; therefore, perform necessary cleanup.
  def rollback_on_create
    # remove the system from elasticsearch
    system_id = "id:#{self.id}"
    Tire::Configuration.client.delete "#{Tire::Configuration.url}/katello_system/_query?q=#{system_id}"
    Tire.index('katello_system').refresh
  end

  private
    def add_pulp_consumer_group record
      record.add_consumers([self.uuid])
    end

    def remove_pulp_consumer_group record
      record.del_consumers([self.uuid])
    end

    def save_task_status pulp_task, task_type, parameters_type, parameters
      TaskStatus.make(self, pulp_task, task_type, parameters_type => parameters)
    end

    def fill_defaults
      self.description = _("Initial Registration Params") unless self.description
      self.location = _("None") unless self.location
    end

    def collect_installed_product_names
      self.installedProducts ? self.installedProducts.map{ |p| p[:productName] } : []
    end

end
